File: AndroidArchitecture/mvp-rxjava/src/main/java/com/codearms/maoqiqi/app/addedittask/AddEditTaskActivity.java
package com.codearms.maoqiqi.app.addedittask;

import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import android.view.View;

import com.codearms.maoqiqi.app.Injection;
import com.codearms.maoqiqi.app.R;
import com.codearms.maoqiqi.app.base.BaseActivity;
import com.codearms.maoqiqi.app.tasks.TasksActivity;

/**
 * Displays an add or edit task screen.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 13:14
 */
public class AddEditTaskActivity extends BaseActivity implements View.OnClickListener {

    private static final String SHOULD_LOAD_DATA_FROM_REPO = "SHOULD_LOAD_DATA_FROM_REPO";

    private AddEditTaskFragment addEditTaskFragment;
    private AddEditTaskPresenter presenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_edit_task);

        String taskId = getIntent().getStringExtra(TasksActivity.EXTRA_TASK_ID);

        findViewById(R.id.fabAddEditTask).setOnClickListener(this);

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle(taskId == null ? R.string.add_task : R.string.edit_task);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null)
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        addEditTaskFragment = (AddEditTaskFragment) getSupportFragmentManager().findFragmentById(R.id.contentFrame);
        if (addEditTaskFragment == null) {
            addEditTaskFragment = AddEditTaskFragment.newInstance();
            getSupportFragmentManager().beginTransaction().add(R.id.contentFrame, addEditTaskFragment).commit();
        }

        boolean shouldLoadDataFromRepo = true;
        // Prevent the presenter from loading data from the repository if this is a config change.
        if (savedInstanceState != null) {
            // Data might not have loaded when the config change happen, so we saved the state.
            shouldLoadDataFromRepo = savedInstanceState.getBoolean(SHOULD_LOAD_DATA_FROM_REPO);
        }

//        presenter = new AddEditTaskPresenter(taskId, Injection.provideTasksRepository(getApplicationContext()),
//                addEditTaskFragment, shouldLoadDataFromRepo, Injection.provideSchedulerProvider());
    }

    @Override
    public void onClick(View v) {
        presenter.addTask(addEditTaskFragment.getTitle(), addEditTaskFragment.getDescription());
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        // Save the state so that next time we know if we need to refresh data.
        outState.putBoolean(SHOULD_LOAD_DATA_FROM_REPO, presenter.isDataMissing());
        super.onSaveInstanceState(outState);
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
}

File: AndroidArchitecture/mvp-kotlin/src/main/java/com/codearms/maoqiqi/app/addedittask/AddEditTaskActivity.kt
package com.codearms.maoqiqi.app.addedittask

import android.os.Bundle
import android.view.View
import com.codearms.maoqiqi.app.Injection
import com.codearms.maoqiqi.app.R
import com.codearms.maoqiqi.app.base.BaseActivity
import com.codearms.maoqiqi.app.tasks.TasksActivity
import com.codearms.maoqiqi.app.utils.add
import com.codearms.maoqiqi.app.utils.setToolbar

/**
 * Displays an add or edit task screen.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 13:14
 */
class AddEditTaskActivity : BaseActivity(), View.OnClickListener {

    private lateinit var addEditTaskFragment: AddEditTaskFragment
    private lateinit var presenter: AddEditTaskPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_edit_task)

        val taskId = intent.getStringExtra(TasksActivity.EXTRA_TASK_ID)

        findViewById<View>(R.id.fabAddEditTask).setOnClickListener(this)

        setToolbar(R.id.toolbar) {
            setTitle(if (taskId == null) R.string.add_task else R.string.edit_task)
            setDisplayHomeAsUpEnabled(true)
        }

        addEditTaskFragment = supportFragmentManager.findFragmentById(R.id.contentFrame) as AddEditTaskFragment?
                ?: AddEditTaskFragment.newInstance().also { add(R.id.contentFrame, it) }

        // Prevent the presenter from loading data from the repository if this is a config change.
        // Data might not have loaded when the config change happen, so we saved the state.
        val shouldLoadDataFromRepo = savedInstanceState?.getBoolean(SHOULD_LOAD_DATA_FROM_REPO)
                ?: true

        presenter = AddEditTaskPresenter(taskId, Injection.provideTasksRepository(applicationContext),
                addEditTaskFragment, shouldLoadDataFromRepo)
    }

    override fun onClick(v: View) {
        presenter.addTask(addEditTaskFragment.title, addEditTaskFragment.description)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        // Save the state so that next time we know if we need to refresh data.
        outState.putBoolean(SHOULD_LOAD_DATA_FROM_REPO, presenter.isDataMissing)
        super.onSaveInstanceState(outState)
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    companion object {
        private const val SHOULD_LOAD_DATA_FROM_REPO = "SHOULD_LOAD_DATA_FROM_REPO"
    }
}